### PR TITLE
Semantic snippets: Work on `ctor` snippet after modifiers

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Originally I wanted to make `ctor` snippet appear after accessibility modifiers and `static` keyword, so user can type e.g. `private ctor` and immediately get private constructor instead of either typing it all himself or changing accessibility modifier after activating `ctor` snippet. Part of that work required to move `GenerateSnippetTextChangeAsync` to `CSharpConstructorSnippetProvider` since `PrecedingModifiers` are only available on C# level. I refactored the method to use more C#-specific APIs to get other things, e.g. used `CSharpSyntaxContext.ContainingTypeDeclaration` instead of searching for node ourselves. As it turned out, these APIs are smarter and handle some edge cases, that were not handled before. So this PR also:
Fixes https://github.com/dotnet/roslyn/issues/68176
Fixes [AB#1818584](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1818584)
Closes https://github.com/dotnet/roslyn/pull/68177 (as a replacement)

### CODE REVIEW

File: src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs [hunk 1]
```diff
@@ -3,25 +3,38 @@
...
 using Microsoft.CodeAnalysis.Snippets.SnippetProviders;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Snippets
 {
     [ExportSnippetProvider(nameof(ISnippetProvider), LanguageNames.CSharp), Shared]
     internal sealed class CSharpConstructorSnippetProvider : AbstractConstructorSnippetProvider
     {
+        private static readonly ISet<SyntaxKind> s_validModifiers = new HashSet<SyntaxKind>(SyntaxFacts.EqualityComparer)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs:29]: 💡 Use `ImmutableSegmentedHashSet<SyntaxKind>`, or even just `ImmutableArray<SyntaxKind>` for one this short

File: src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs [hunk 2]
```diff
@@ -33,13 +46,41 @@ protected override async Task<bool> IsValidSnippetLocationAsync(Document documen
             var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
             var syntaxContext = (CSharpSyntaxContext)document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
 
+            var precedingModifiers = syntaxContext.PrecedingModifiers;
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs:49]: 📝 This is going to conflict with a fairly significant rewrite I just completed in this space. I'm thinking we maybe wait until after that merges.

File: src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs:29]: The pattern of declaring `s_validModifiers` as `new HashSet<SyntaxKind>(SyntaxFacts.EqualityComparer)` is very common in roslyn, e.g. all C# keyword recommenders use it ([example](https://github.com/dotnet/roslyn/blob/60eaadf5c5e8782e43ed2455c80380d6243eac8b/src/Features/CSharp/Portable/Completion/KeywordRecommenders/UnsafeKeywordRecommender.cs#L14-L61)). So I think if we change it, we should chage it everywhere at once, which is better to be done in a separate PR

File: src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs (hunk 2 above)
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs:49]: Well, I couldn't predic that. Please ping me here when your changes are merged. For now I'll make this PR a draft

Comment (sharwell) [reply src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs:49]: If I don't reply back by tomorrow, feel free to ping me or make this ready for review again.

**Comment by DoctorKrolic** [comment]
@sharwell What are we gonna do with this PR?

**Comment by sharwell** [comment]
@DoctorKrolic My PR to fix this is going to be in #71872. I was delayed by the unexpected failure in the integration test pipeline but should be able to move it forward over the next few hours.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpConstructorSnippetCompletionProviderTests.cs [hunk 3]
```diff
@@ -254,5 +254,132 @@ public MyClass1()
...
+        [InlineData("override")]
+        [InlineData("readonly")]
+        [InlineData("new")]
+        [InlineData("file")]
+        public async Task ConstructorSnippetMissingAfterInvalidModifierTest(string modifier)
+        {
+            var markupBeforeCommit = $$"""
+                class MyClass
+                {
+                    {{modifier}} $$
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpConstructorSnippetCompletionProviderTests.cs:301]: 💡 Can you add tests covering cases where a constructor is being added before other members, e.g.:

```
class C
{
  public $$
  readonly int Value = 3;
}
```

Comment (DoctorKrolic) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpConstructorSnippetCompletionProviderTests.cs:301]: Added 3 variations of this case

**Review by akhera99 — APPROVED** [review]

**Comment by akhera99** [comment]
Looks good, thanks for the contribution @DoctorKrolic!

**Pull request merged**